### PR TITLE
[Windows] Ensure files open with FILE_SHARE_DELETE

### PR DESCRIPTION
**Pull request by gmittert**
On Windows, open() doesn't set FILE_SHARE_DELETE by default meaning that
if another handle requests deletion privileges, it will always fail
unless the original handle is closed.

An example of where this happens is
```swift
{
    guard let fh = FileHandle(forWritingAtPath: somePath) else { ... }
    defer { try! FileManager.default.removeItem(atPath: somePath) }
    fh.write()
}
```
Without FILE_SHARE_DELETE, the defer happens before the FileHandle's deinit occurs, causing removeItem to fail because it can't open somePath for deletion.

### CODE REVIEW

**Comment by gmittert** [comment]
@millenomi @compnerd

File: CoreFoundation/Base.subproj/CFPlatform.c [hunk 1]
```diff
@@ -884,11 +884,15 @@ CF_EXPORT int _NS_rename(const char *oldName, const char *newName) {
...
 CF_EXPORT int _NS_open(const char *name, int oflag, int pmode) {
+#if TARGET_OS_WIN32
+    return _CFOpenFileWithMode(name, oflag, pmode);
+#else
     wchar_t *wide = createWideFileSystemRepresentation(name, NULL);
     int fd;
     _wsopen_s(&fd, wide, oflag, _SH_DENYNO, _S_IREAD | _S_IWRITE);
     free(wide);
     return fd;
+#endif
```
Comment (compnerd) [inline CoreFoundation/Base.subproj/CFPlatform.c]: This code is Windows specific already, so this change doesn't make sense.  Just kill the existing implementation.  Although, I think that we want to inline the definition that you placed in `_CFOpenFileWithMode` here as `open` is being replaced with `_NS_open` on Windows.

File: CoreFoundation/Base.subproj/CFPlatform.c [hunk 2]
```diff
@@ -1444,10 +1448,85 @@ CF_EXPORT char **_CFEnviron(void) {
...
+        break;
+      case _O_WRONLY:
+        dwDesiredAccess = GENERIC_WRITE;
+        break;
+      case _O_RDWR:
+        dwDesiredAccess = GENERIC_READ | GENERIC_WRITE;
+        break;
+    }
+
+    DWORD dwCreationDisposition = OPEN_EXISTING;
```
Comment (compnerd) [inline CoreFoundation/Base.subproj/CFPlatform.c]: What happens if `_O_CREAT` was passed?  I think it is better to move L1475-L1477 here.

File: CoreFoundation/Base.subproj/CFPlatform.c [hunk 3]
```diff
@@ -1444,10 +1448,85 @@ CF_EXPORT char **_CFEnviron(void) {
...
+          errno = ENOENT;
+          break;
+        default:
+          errno = EIO;
+          break;
+      }
+      return -1;
+    }
+
+    // _open_osfhandle handles _O_APPEND and _ORDONLY
```
Comment (compnerd) [inline CoreFoundation/Base.subproj/CFPlatform.c]: Nit: missing `_` in `_O_RDONLY`.

File: CoreFoundation/Base.subproj/CFPlatform.c [hunk 4]
```diff
@@ -885,9 +885,73 @@ CF_EXPORT int _NS_rename(const char *oldName, const char *newName) {
...
+    }
+
+    // Backup semantics are required to receive a handle to a directory
+    DWORD dwFlagsAndAttributes = FILE_FLAG_BACKUP_SEMANTICS;
+    if (pmode & _S_IREAD) {
+      dwFlagsAndAttributes |= FILE_ATTRIBUTE_READONLY;
+    }
+
+    HANDLE handle = CreateFileW(wide, dwDesiredAccess, FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE,
+        /* lpSecurityAttributes= */nil, dwCreationDisposition, dwFlagsAndAttributes,
```
Comment (compnerd) [inline CoreFoundation/Base.subproj/CFPlatform.c]: s/nil/NULL/

File: CoreFoundation/Base.subproj/CFPlatform.c [hunk 5]
```diff
@@ -885,9 +885,73 @@ CF_EXPORT int _NS_rename(const char *oldName, const char *newName) {
...
+        /* hTemplatefile= */ nil);
```
Comment (compnerd) [inline CoreFoundation/Base.subproj/CFPlatform.c]: s/nil/NULL/

File: CoreFoundation/Base.subproj/CFPlatform.c [hunk 6]
```diff
@@ -1446,8 +1510,13 @@ CF_EXPORT char **_CFEnviron(void) {
 CF_CROSS_PLATFORM_EXPORT int _CFOpenFileWithMode(const char *path, int opts, mode_t mode) {
     return open(path, opts, mode);
 }
+
 int _CFOpenFile(const char *path, int opts) {
+#if TARGET_OS_WIN32
+    return open(path, opts, 0);
```
Comment (compnerd) [inline CoreFoundation/Base.subproj/CFPlatform.c]: Just use the 3 parameter `open` always.  Its equivalent.

**Comment by gmittert** [comment]
@swift-ci please test

**Comment by gmittert** [comment]
@swift-ci please test

**Pull request merged**